proposals/simd/portable-simd.md

```diff
@@ -0,0 +1,1320 @@
+    result = S.BoolType.New()
+    for i in range(S.Lanes):
+        result[i] = func(a[i], b[i])
+    return result
+```
+
+## Constructing SIMD values
+
+### Build vector from individual lanes
+* `b8x16.build(x: boolean[16]) -> b8x16`
+* `b16x8.build(x: boolean[8]) -> b16x8`
+* `b32x4.build(x: boolean[4]) -> b32x4`
+* `b64x2.build(x: boolean[2]) -> b64x2`
+* `i8x16.build(x: i8[16]) -> v128`
+* `i16x8.build(x: i16[8]) -> v128`
+* `i32x4.build(x: i32[4]) -> v128`
+* `i64x2.build(x: i64[2]) -> v128`
+* `f32x4.build(x: f32[4]) -> v128`
+* `f64x2.build(x: f64[2]) -> v128`
+
+Construct a vector from an array of individual lane values.
+
+```python
+def S.build(x):
+    result = S.New()
+    for i in range(S.Lanes):
+        result[i] = x[i]
+    return result
+```
+
+### Create vector with identical lanes
+* `b8x16.splat(x: boolean) -> b8x16`
+* `b16x8.splat(x: boolean) -> b16x8`
+* `b32x4.splat(x: boolean) -> b32x4`
+* `b64x2.splat(x: boolean) -> b64x2`
+* `i8x16.splat(x: i8) -> v128`
+* `i16x8.splat(x: i16) -> v128`
+* `i32x4.splat(x: i32) -> v128`
+* `i64x2.splat(x: i64) -> v128`
+* `f32x4.splat(x: f32) -> v128`
+* `f64x2.splat(x: f64) -> v128`
+
+Construct a vector with `x` replicated to all lanes:
+
+```python
+def S.splat(x):
+    result = S.New()
+    for i in range(S.Lanes):
+        result[i] = x
+    return result
+```
+
+## Accessing lanes
+
+### Extract lane as a scalar
+* `b8x16.extractLane(a: b8x16, i: LaneIdx16) -> boolean`
+* `b16x8.extractLane(a: b16x8, i: LaneIdx8) -> boolean`
+* `b32x4.extractLane(a: b32x4, i: LaneIdx4) -> boolean`
+* `b64x2.extractLane(a: b64x2, i: LaneIdx2) -> boolean`
+* `i8x16.extractLane(a: v128, i: LaneIdx16) -> i8`
+* `i16x8.extractLane(a: v128, i: LaneIdx8) -> i16`
+* `i32x4.extractLane(a: v128, i: LaneIdx4) -> i32`
+* `i64x2.extractLane(a: v128, i: LaneIdx2) -> i64`
+* `f32x4.extractLane(a: v128, i: LaneIdx4) -> f32`
+* `f64x2.extractLane(a: v128, i: LaneIdx2) -> f64`
+
+Extract the value of lane `i` in `a`.
+
+```python
+def S.extractLane(a, i):
+    return a[i]
+```
+
+### Replace lane value
+* `b8x16.replaceLane(a: b8x16, i: LaneIdx16, x: boolean) -> b8x16`
+* `b16x8.replaceLane(a: b16x8, i: LaneIdx8, x: boolean) -> b16x8`
+* `b32x4.replaceLane(a: b32x4, i: LaneIdx4, x: boolean) -> b32x4`
+* `b64x2.replaceLane(a: b64x2, i: LaneIdx2, x: boolean) -> b64x2`
+* `i8x16.replaceLane(a: v128, i: LaneIdx16, x: i8) -> v128`
+* `i16x8.replaceLane(a: v128, i: LaneIdx8, x: i16) -> v128`
+* `i32x4.replaceLane(a: v128, i: LaneIdx4, x: i32) -> v128`
+* `i64x2.replaceLane(a: v128, i: LaneIdx2, x: i64) -> v128`
+* `f32x4.replaceLane(a: v128, i: LaneIdx4, x: f32) -> v128`
+* `f64x2.replaceLane(a: v128, i: LaneIdx2, x: f64) -> v128`
+
+Return a new vector with lanes identical to `a`, except for lane `i` which has
+the value `x`.
+
+```python
+def S.replaceLane(a, i, x):
+    result = S.New()
+    for j in range(S.Lanes):
+        result[j] = a[j]
+    result[i] = x
+    return result
```

Comment:
My intention was for these indices to be immediate operands. Do you think we need dynamic lane indices for these operations? It's hard to generate good code for that.